src/main/java/io/confluent/connect/jdbc/source/TimestampIncrementingTableQuerier.java

```diff
@@ -229,6 +202,56 @@ public SourceRecord extractRecord() throws SQLException {
     return new SourceRecord(partition, offset.toMap(), topic, record.schema(), record);
   }
 
+  // Visible for testing
+  TimestampIncrementingOffset extractOffset(Schema schema, Struct record) {
+    final Timestamp extractedTimestamp;
+    if (timestampColumn != null) {
+      extractedTimestamp = (Timestamp) record.get(timestampColumn);
+      Timestamp timestampOffset = offset.getTimestampOffset();
+      assert timestampOffset != null && timestampOffset.compareTo(extractedTimestamp) <= 0;
+    } else {
+      extractedTimestamp = null;
+    }
+
+    final Long extractedId;
+    if (incrementingColumn != null) {
+      final Schema incrementingColumnSchema = schema.field(incrementingColumn).schema();
+      final Object incrementingColumnValue = record.get(incrementingColumn);
+      if (incrementingColumnValue == null) {
+        throw new ConnectException("Null value for incrementing column of type: " + incrementingColumnSchema.type());
+      } else if (isIntegralPrimitiveType(incrementingColumnValue)) {
+        extractedId = ((Number) incrementingColumnValue).longValue();
+      } else if (incrementingColumnSchema.name() != null && incrementingColumnSchema.name().equals(Decimal.LOGICAL_NAME)) {
+        final BigDecimal decimal = ((BigDecimal) incrementingColumnValue);
+        if (decimal.compareTo(LONG_MAX_VALUE_AS_BIGDEC) > 0) {
+          throw new ConnectException("Decimal value for incrementing column exceeded Long.MAX_VALUE");
+        }
+        if (decimal.scale() != 0) {
```

Comment:
> so I'll give this another go.

Er, re-abandoned. The issue is, if there is no existing stored offset, you have a `null` but no type. While a sensible default value can be determined ("" for string, -1 for numbers..), need type-info at bind-time.

I think happy with the solution for now. We could expand support to all `Number` types rather than just upto `Long.MAX_VALUE` if there is a request for it.